File: src/collect.mjs
/**
 * Collect the prompts *you* typed into Claude Code on a given day.
 *
 * Reads only the transcripts Claude Code keeps on your own machine
 * (`~/.claude/projects/<project>/<session>.jsonl`). Nothing is uploaded and no
 * network call is made.
 *
 * What counts as "a prompt you typed":
 *   `type === "user"` records, minus everything the machine injected — tool results,
 *   `isMeta`, pure `<system-reminder>` lines, `<local-command-stdout>`, agent-to-agent
 *   messages, task notifications, bash I/O, continuation summaries, interrupts and
 *   sub-agent turns. Slash commands are counted separately.
 *
 * Session naming falls back in this order, so it works on any machine:
 *   1. `agentName` / `customTitle` recorded in the transcript
 *   2. a `sessionLabel` rule from the config file
 *   3. the first 8 characters of the session UUID
 */
import { createReadStream } from "node:fs";
import { readdir, readFile } from "node:fs/promises";
import { createInterface } from "node:readline";
import path from "node:path";
import os from "node:os";

export const DEFAULT_PROJECTS_DIR = path.join(os.homedir(), ".claude", "projects");

const pad = (n) => String(n).padStart(2, "0");

/** Split an ISO timestamp into local (or fixed-offset) date and time parts. */
export function localParts(iso, tzOffsetHours) {
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return null;
  if (tzOffsetHours == null) {
    return {
      date: `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`,
      time: `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`,
    };
  }
  const shifted = new Date(d.getTime() + tzOffsetHours * 3600 * 1000);
  return {
    date: `${shifted.getUTCFullYear()}-${pad(shifted.getUTCMonth() + 1)}-${pad(shifted.getUTCDate())}`,
    time: `${pad(shifted.getUTCHours())}:${pad(shifted.getUTCMinutes())}:${pad(shifted.getUTCSeconds())}`,
  };
}

/** Today's date in the target timezone. */
export function todayIn(tzOffsetHours) {
  return localParts(new Date().toISOString(), tzOffsetHours).date;
}

/**
 * UTC dates that can hold records belonging to one local day. A local day never
 * reaches further than the day before or after in UTC, so these three prefixes
 * are enough to pre-filter lines cheaply before parsing them.
 */
function candidateDates(day) {
  const base = new Date(`${day}T12:00:00Z`).getTime();
  return [-1, 0, 1].map((offset) => new Date(base + offset * 86400000).toISOString().slice(0, 10));
}

export function classify(text) {
  if (text.includes("Another Claude session sent a message") || text.includes("<teammate-message")) return "teammate";
  if (text.startsWith("<task-notification>")) return "task_notif";
  if (text.startsWith("<bash-input>") || text.startsWith("<bash-stdout>") || text.startsWith("<bash-stderr>")) return "bash_io";
  if (text.startsWith("This session is being continued")) return "continuation";
  if (text.startsWith("[Request interrupted")) return "interrupt";
  if (text.includes("<command-name>") || text.includes("<command-message>")) return "slash";
  return "human";
}

/** Pull the human-typed body out of a user record, or null if it is not one. */
export function userText(obj) {
  if (obj.type !== "user" || obj.isMeta || obj.isSidechain) return null;
  const content = obj.message?.content;
  let text;
  if (typeof content === "string") {
    text = content;
  } else if (Array.isArray(content)) {
    // A tool_result anywhere means this is a tool's return value, not something a person typed.
    for (const item of content) if (item?.type === "tool_result") return null;
    text = content.filter((i) => i?.type === "text").map((i) => i.text ?? "").join("\n");
  } else {
    return null;
  }
  text = text.trim();
  if (!text || text.includes("<local-command-stdout>") || text.startsWith("Caveat:")) return null;
  if (text.startsWith("<system-reminder>") && text.endsWith("</system-reminder>")) return null;
  return text;
}

export async function loadConfig(configPath) {
  if (!configPath) return {};
  try {
    return JSON.parse(await readFile(configPath, "utf8")) ?? {};
  } catch (err) {
    if (err.code === "ENOENT") return {};
    throw new Error(`config could not be read (${configPath}): ${err.message}`);
  }
}

async function listTranscripts(root) {
  let entries;
  try {
    entries = await readdir(root, { withFileTypes: true });
  } catch (err) {
    if (err.code === "ENOENT") return [];
    throw err;
  }
  const files = [];
  for (const entry of entries) {
    if (!entry.isDirectory() || entry.name === "subagents") continue;
    const dir = path.join(root, entry.name);
    for (const name of await readdir(dir)) {
      if (name.endsWith(".jsonl")) files.push(path.join(dir, name));
    }
  }
  return files.sort();
}

/** Read one transcript, returning the day's records plus whatever names it carries. */
async function scanFile(file, day, prefixes, tz, labelPattern) {
  const rows = [];
  const meta = { cwd: null, agentName: null, sessionId: null, ruleHits: new Map(), file };
  const cheapKeys = prefixes.map((p) => `"${p}`);

  const rl = createInterface({
    input: createReadStream(file, { encoding: "utf8" }),
    crlfDelay: Infinity,
  });
  try {
    for await (const line of rl) {
      if (!line) continue;
      const mayHaveDate = cheapKeys.some((key) => line.includes(key));
      const mayHaveName = line.includes('"agent-name"') || line.includes('"custom-title"');
      if (!mayHaveDate && !mayHaveName) continue;

      let obj;
      try {
        obj = JSON.parse(line);
      } catch {
        continue;
      }
      meta.sessionId ??= obj.sessionId ?? null;
      if (obj.type === "agent-name" && obj.agentName) { meta.agentName ??= obj.agentName; continue; }
      if (obj.type === "custom-title" && obj.customTitle) { meta.agentName ??= obj.customTitle; continue; }
      if (!mayHaveDate || !obj.timestamp) continue;

      // Filter on the *local* date, not the raw UTC prefix, so "a day" is the day you lived.
      const parts = localParts(obj.timestamp, tz);
      if (!parts || parts.date !== day) continue;
      meta.cwd ??= obj.cwd ?? null;

      if (labelPattern) {
        for (const match of line.matchAll(labelPattern)) {
          const captured = match[1] ?? match[0];
          meta.ruleHits.set(captured, (meta.ruleHits.get(captured) ?? 0) + 1);
        }
      }
      const text = userText(obj);
      if (text === null) continue;
      rows.push({
        time: parts.time,
        text,
        promptSource: obj.promptSource ?? null,
        branch: obj.gitBranch ?? null,
      });
    }
  } finally {
    rl.close();
  }
  return { rows, meta };
}

function sessionLabel(meta, cfg) {
  if (meta.agentName) return meta.agentName;
  const names = cfg.sessionLabel?.names ?? {};
  if (meta.ruleHits.size) {
    const [captured] = [...meta.ruleHits.entries()].sort((a, b) => b[1] - a[1])[0];
    if (captured in names) return names[captured];
    if (captured) return String(captured);
  }
  const id = meta.sessionId ?? path.basename(meta.file, ".jsonl");
  return id ? id.slice(0, 8) : "unknown";
}

function workspaceLabel(meta, cfg) {
  if (!meta.cwd) return "";
  const base = path.basename(meta.cwd.replace(/\/+$/, ""));
  return cfg.workspaceAlias?.[base] ?? base;
}

export async function collect(options = {}) {
  const {
    date,
    projectsDir = DEFAULT_PROJECTS_DIR,
    config = {},
    tz = null,
    includeText = true,
    dropSdk = false,
  } = options;

  const day = date ?? todayIn(tz);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(day)) throw new Error(`date must look like YYYY-MM-DD, got: ${day}`);

  const root = projectsDir.startsWith("~")
    ? path.join(os.homedir(), projectsDir.slice(1))
    : projectsDir;
  const files = await listTranscripts(root);
  if (!files.length) throw new Error(`no transcripts found under ${root}`);

  const labelPattern = config.sessionLabel?.pattern ? new RegExp(config.sessionLabel.pattern, "g") : null;
  const prefixes = candidateDates(day);

  const seen = new Set();
  const counts = new Map();
  const human = [];
  const slash = [];
  const bump = (key) => counts.set(key, (counts.get(key) ?? 0) + 1);

  for (const file of files) {
    const { rows, meta } = await scanFile(file, day, prefixes, tz, labelPattern);
    if (!rows.length) continue;
    const agent = sessionLabel(meta, config);
    const workspace = workspaceLabel(meta, config);
    const session = workspace ? `${workspace}.${agent}` : agent;

    for (const row of rows) {
      // The same conversation can show up in several files (resume / mirrored sessions).
      const key = `${row.time}\n${row.text.slice(0, 80)}`;
      if (seen.has(key)) continue;
      seen.add(key);

      const kind = classify(row.text);
      bump(kind);
      if (dropSdk && (row.promptSource === "sdk" || row.promptSource === "system")) {
        bump("sdk_filtered");
        continue;
      }
      if (kind === "human") {
        const record = { time: row.time, session, workspace, agent };
        if (includeText) record.text = row.text;
        record.chars = row.text.length;
        if (row.branch) record.branch = row.branch;
        human.push(record);
      } else if (kind === "slash") {
        const m = row.text.match(/<command-name>\s*([^<]+)/) ?? row.text.match(/<command-message>\s*([^<]+)/);
        slash.push({ time: row.time, session, workspace, agent, command: m ? m[1].trim() : "?" });
      }
    }
  }

  human.sort((a, b) => a.time.localeCompare(b.time));
  slash.sort((a, b) => a.time.localeCompare(b.time));

  const bySession = {};
  for (const row of human) bySession[row.session] = (bySession[row.session] ?? 0) + 1;

  return {
    date: day,
    generatedAt: new Date().toISOString(),
    note: "Prompts typed by a human. Mirrored sessions de-duplicated; machine-injected turns excluded. session = <workspace>.<agent>.",
    summary: {
      human_prompts: human.length,
      slash_commands: slash.length,
      by_session: Object.fromEntries(Object.entries(bySession).sort((a, b) => b[1] - a[1])),
      excluded: Object.fromEntries(
        ["teammate", "task_notif", "bash_io", "continuation", "interrupt", "sdk_filtered"]
          .map((key) => [key, counts.get(key) ?? 0]),
      ),
    },
    agents: config.agents ?? {},
    human_prompts: human,
    slash_commands: slash,
  };
}

File: bin/collect.py
#!/usr/bin/env python3
"""Claude Code のローカル transcript から「その日に人が打ったプロンプト」を集めて JSON にする。

使い方:
    python3 bin/collect.py --date 2026-09-16 --out out/2026-09-16.json

読むのは `~/.claude/projects/*/*.jsonl`（Claude Code が自分のマシンに残す会話ログ）だけ。
ネットワークへは一切出ない。

集計方針:
  - `type == "user"` のうち、**人が打ったもの**だけを human として残す。
    tool_result / isMeta / <system-reminder> だけの行 / <local-command-stdout> / Caveat は捨てる。
  - サブエージェント（isSidechain）は別セッション扱いにせず除外する。
  - 同じ会話が複数ファイルに写っていることがある（resume・ミラー）ので
    (時刻[秒], 本文先頭 80 字) で重複を除く。
  - エージェント間連携・task 通知・bash 入出力・継続サマリ・中断通知は human から外し、
    件数だけ summary.excluded に残す。
  - スラッシュコマンドは slash_commands に分ける。

セッション名の決め方（環境に依存しない順で落ちていく）:
  1. transcript の `agentName` / `customTitle`（Claude Code が持っている名前）
  2. --config で渡した JSON の `sessionLabel` ルール（正規表現で拾って名前に対応づける）
  3. セッション UUID の先頭 8 桁
ワークスペース名は transcript の `cwd` の末尾ディレクトリ名。config の `workspaceAlias` で短くできる。
"""
import argparse, glob, json, os, re, sys
from collections import Counter
from datetime import datetime, timezone, timedelta

DEFAULT_PROJECTS_DIR = "~/.claude/projects"


def parse_args(argv):
    p = argparse.ArgumentParser(description="Claude Code の transcript から当日のプロンプトを集める")
    p.add_argument("--date", help="対象日 YYYY-MM-DD（既定: 今日）")
    p.add_argument("--out", help="出力 JSON パス（既定: out/<date>.json）")
    p.add_argument("--projects-dir", default=DEFAULT_PROJECTS_DIR,
                   help=f"transcript の置き場（既定: {DEFAULT_PROJECTS_DIR}）")
    p.add_argument("--config", help="セッション名・色の設定 JSON（既定: config/agents.json があれば使う）")
    p.add_argument("--tz", type=float, help="表示タイムゾーンの UTC オフセット時間（既定: このマシンの設定）")
    p.add_argument("--include-text", dest="include_text", action="store_true", default=True,
                   help="プロンプト本文を出力に含める（既定）")
    p.add_argument("--no-text", dest="include_text", action="store_false",
                   help="本文を伏せて時刻・セッションだけ出す（人に見せる用）")
    p.add_argument("--drop-sdk", action="store_true",
                   help="promptSource が sdk/system のものも落とす（自動実行を強めに除く）")
    return p.parse_args(argv)


def load_config(path):
    """設定ファイルを読む。無ければ空（＝素のまま動く）。"""
    if not path:
        here = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "config", "agents.json")
        path = here if os.path.exists(here) else None
    if not path or not os.path.exists(path):
        return {}
    with open(path, encoding="utf-8") as fh:
        return json.load(fh) or {}


def tzinfo_for(offset_hours):
    if offset_hours is None:
        return datetime.now().astimezone().tzinfo
    return timezone(timedelta(hours=offset_hours))


def classify(text):
    """人が打ったものか、機械が差し込んだものかを分ける。"""
    if "Another Claude session sent a message" in text or "<teammate-message" in text:
        return "teammate"
    if text.startswith("<task-notification>"):
        return "task_notif"
    if text.startswith(("<bash-input>", "<bash-stdout>", "<bash-stderr>")):
        return "bash_io"
    if text.startswith("This session is being continued"):
        return "continuation"
    if text.startswith("[Request interrupted"):
        return "interrupt"
    if "<command-name>" in text or "<command-message>" in text:
        return "slash"
    return "human"


def user_text(obj):
    """user レコードから人が打った本文を取り出す。対象外なら None。"""
    if obj.get("type") != "user" or obj.get("isMeta") or obj.get("isSidechain"):
        return None
    message = obj.get("message") or {}
    content = message.get("content")
    if isinstance(content, str):
        text = content
    elif isinstance(content, list):
        # tool_result が 1 つでも混ざっていればツールの戻り値であって、人の発話ではない
        for item in content:
            if isinstance(item, dict) and item.get("type") == "tool_result":
                return None
        text = "\n".join(i.get("text", "") for i in content if isinstance(i, dict) and i.get("type") == "text")
    else:
        return None
    text = text.strip()
    if not text or "<local-command-stdout>" in text or text.startswith("Caveat:"):
        return None
    if text.startswith("<system-reminder>") and text.endswith("</system-reminder>"):
        return None
    return text


def scan_file(path, day, label_rules):
    """1 ファイルを読んで (レコード列, セッション名, ワークスペース名) を返す。"""
    rows, cwd, agent_name, session_id = [], None, None, None
    rule_hits = Counter()
    try:
        fh = open(path, encoding="utf-8")
    except OSError:
        return [], None, None
    with fh:
        for line in fh:
            # 当日分が 1 行も無いファイルでも、名前を決めるヘッダ行は拾いたいので
            # 安い文字列判定で本文パースを間引きつつ、メタ行だけは常に見る。
            cheap_skip = f'"{day}' not in line
            if cheap_skip and '"agent-name"' not in line and '"custom-title"' not in line:
                continue
            try:
                obj = json.loads(line)
            except ValueError:
                continue
            rtype = obj.get("type")
            session_id = session_id or obj.get("sessionId")
            if rtype == "agent-name" and obj.get("agentName"):
                agent_name = agent_name or obj["agentName"]
                continue
            if rtype == "custom-title" and obj.get("customTitle"):
                agent_name = agent_name or obj["customTitle"]
                continue
            if cheap_skip:
                continue
            ts = obj.get("timestamp", "")
            if not ts.startswith(day):
                continue
            cwd = cwd or obj.get("cwd")
            # 設定の正規表現ルール（例: 自分の運用で使っているエージェント番号を拾う）
            for name, pattern in label_rules:
                for m in pattern.findall(line):
                    rule_hits[(name, m if isinstance(m, str) else m[0])] += 1
            text = user_text(obj)
            if text is None:
                continue
            rows.append({
                "ts": ts,
                "text": text,
                "promptSource": obj.get("promptSource"),
                "gitBranch": obj.get("gitBranch"),
            })
    return rows, {"cwd": cwd, "agent_name": agent_name, "session_id": session_id,
                  "rule_hits": rule_hits, "path": path}, None


def session_label(meta, cfg):
    """セッションの表示名を決める。環境依存の推測はせず、取れたものから順に使う。"""
    if meta.get("agent_name"):
        return meta["agent_name"]
    rules = cfg.get("sessionLabel") or {}
    names = rules.get("names") or {}
    hits = meta.get("rule_hits") or Counter()
    if hits:
        (_, captured), _ = hits.most_common(1)[0]
        if captured in names:
            return names[captured]
        if captured:
            return str(captured)
    sid = meta.get("session_id") or os.path.basename(meta.get("path", ""))
    return sid[:8] if sid else "unknown"


def workspace_label(meta, cfg):
    cwd = meta.get("cwd")
    if not cwd:
        return ""
    base = os.path.basename(cwd.rstrip("/"))
    return (cfg.get("workspaceAlias") or {}).get(base, base)


def main(argv=None):
    args = parse_args(argv if argv is not None else sys.argv[1:])
    cfg = load_config(args.config)
    tz = tzinfo_for(args.tz)
    day = args.date or datetime.now(tz).strftime("%Y-%m-%d")
    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", day):
        sys.exit(f"--date は YYYY-MM-DD 形式で: {day!r}")
    out = args.out or os.path.join("out", f"{day}.json")

    label_rules = []
    rules = (cfg.get("sessionLabel") or {}).get("pattern")
    if rules:
        label_rules.append(("pattern", re.compile(rules)))

    root = os.path.expanduser(args.projects_dir)
    files = [f for f in glob.glob(os.path.join(root, "*", "*.jsonl")) if "/subagents/" not in f]
    if not files:
        sys.exit(f"transcript が見つからない: {root}/*/*.jsonl")

    seen, counts = set(), Counter()
    human, slash = [], []
    for path in files:
        rows, meta, _ = scan_file(path, day, label_rules)
        if not rows:
            continue
        name = session_label(meta, cfg)
        ws = workspace_label(meta, cfg)
        session = f"{ws}.{name}" if ws else name
        for row in rows:
            try:
                local = datetime.fromisoformat(row["ts"].replace("Z", "+00:00")).astimezone(tz)
            except ValueError:
                continue
            hhmmss = local.strftime("%H:%M:%S")
            key = (hhmmss, row["text"][:80])
            if key in seen:
                continue
            seen.add(key)
            kind = classify(row["text"])
            counts[kind] += 1
            if args.drop_sdk and row.get("promptSource") in ("sdk", "system"):
                counts["sdk_filtered"] += 1
                continue
            if kind == "human":
                rec = {"time": hhmmss, "session": session, "workspace": ws, "agent": name}
                if args.include_text:
                    rec["text"] = row["text"]
                rec["chars"] = len(row["text"])
                if row.get("gitBranch"):
                    rec["branch"] = row["gitBranch"]
                human.append(rec)
            elif kind == "slash":
                m = re.search(r"<command-name>\s*([^<]+)", row["text"]) or \
                    re.search(r"<command-message>\s*([^<]+)", row["text"])
                slash.append({"time": hhmmss, "session": session, "workspace": ws, "agent": name,
                              "command": m.group(1).strip() if m else "?"})

    human.sort(key=lambda r: r["time"])
    slash.sort(key=lambda r: r["time"])
    doc = {
        "date": day,
        "generatedAt": datetime.now(tz).isoformat(timespec="seconds"),
        "note": "人が打ったプロンプトの記録。重複セッション除去・自動投入分は除外。session=<workspace>.<agent>。",
        "summary": {
            "human_prompts": len(human),
            "slash_commands": len(slash),
            "by_session": dict(Counter(r["session"] for r in human).most_common()),
            "excluded": {k: counts.get(k, 0) for k in
                         ("teammate", "task_notif", "bash_io", "continuation", "interrupt", "sdk_filtered")},
        },
        "agents": cfg.get("agents") or {},
        "human_prompts": human,
        "slash_commands": slash,
    }
    os.makedirs(os.path.dirname(os.path.abspath(out)), exist_ok=True)
    with open(out, "w", encoding="utf-8") as fh:
        json.dump(doc, fh, ensure_ascii=False, indent=2)
    print(f"saved: {out}")
    print(f"human={len(human)} slash={len(slash)} sessions={len(doc['summary']['by_session'])}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
